Add new tests to cover some corner cases

ricardotulio opened the pull request:
## Description
This pull request add tests to cover some error scenarios.

## Changes
Add tests to validate when:
- Try to create a user without a username
- Try to create a user with an invalid username
- Try to get a user that doesn't exists

ricardotulio committed:
Add new tests to cover exception scenarios

Add tests that validate when
- username doesnt exists
- create user using a invalid username
- create user without a username

mverardo commented on src/main/java/com/ministore/api/controllers/GetUserHandler.java:
```diff
@@ -23,7 +26,13 @@ public GetUserHandler(UserRepository repository) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "entity not found");
+        }
+
+        return user;
```
There is an interesting case here.
`this.repository.findById()` returns an `Option<User>`, but if you are always checking if the `user.isEmpty` before returning, then `getUser()` doesn't really return an `Option<User>` anymore. It either returns an `User` or throws the exception.

It is important to change the return type of `getUser()` to `User`, because this signals to callers of this function that they don't need to deal with the case where this function returns an empty `Option`.

ricardotulio replied:
Awesome! I applied your suggestion from the previous comment about using `.orElseThrow` and implemented the same approach here.

mverardo commented on src/main/java/com/ministore/api/controllers/GetUserHandler.java:
```diff
@@ -23,7 +26,13 @@ public GetUserHandler(UserRepository repository) {
-        return this.repository.findById(id);
+        Optional<User> user = this.repository.findById(id);
+
+        if (user.isEmpty()) { 
```
A good alternative would be to remove this `if` check and use the [`user.orElseThrow()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElseThrow-java.util.function.Supplier-) method instead.

Let me know if you have any questions on how to do it.

mverardo commented on src/test/java/com/ministore/api/integration/CreateUserTest.java:
```diff
@@ -51,4 +51,39 @@ void mustBeAbleToCreateAUser() throws Exception
+            .content("{}")
+            .accept(MediaType.APPLICATION_JSON))
+            .andExpect(status().isBadRequest());
+    }
```
Nice tests!

I'd like to ask you to think about two new cases, and add tests accordingly to cement the behavior:

* What should happen when the same username is sent more than once? i.e. what does the system do if we make the `POST` request with body `"{\"username\":\"ricardotulio\"}"` twice?
* What should happen when the request body includes more fields than the username? i.e.: what does the system do if we receive a body like this: `"{\"username\":\"some.username\", password: \"some password\"}"`?

mverardo replied:
Following our conversation today, can you add a test that checks the second case?

mverardo reviewed (changes requested):
Nice PR! 🔥 

Besides the points I've added on other comments, It might be a good idea for us to talk about how the database is being managed: did you run simple scripts directly on the database? Or are you using some kind of Migration tool to create tables for you?

mverardo commented on src/test/java/com/ministore/api/integration/CreateUserTest.java:
```diff
@@ -51,4 +51,39 @@ void mustBeAbleToCreateAUser() throws Exception
+    @Test
+    void mustReturnBadRequestWhenUsernameIsInvalid() throws Exception {
+        this.mockMvc.perform(
+            post("/user")
```
Ah, just another nitpick I should have seen earlier:

Rest URIs represent collections of resources on the server, so they should be written as plural. So the right URI here would be `/users` instead of `/user`.

ricardotulio committed:
Ensure no duplicate usernames

Ensure that the application doesn't allow to add two equal usernames.

ricardotulio committed:
Fix resource path: change the endpoint form /user to /users

ricardotulio committed:
Refact GetUserHandler to user .orElseThrow

ricardotulio committed:
Refact CreateUserTest and remove duplicated code

ricardotulio committed:
Test: Ensure that the username returned is equal to the username sent

Ensure that the username returned from a post to /users is exactly the
same sent in the request.

ricardotulio commented on src/main/java/com/ministore/api/controllers/CreateUserHandler.java:
```diff
@@ -32,7 +33,15 @@ private User createUser(@RequestBody CreateUserRequest request)
+        try {
+            user = this.userService.createUser(request.username);
+        } catch(DataIntegrityViolationException exception) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "username is invalid");
```
Some comments:
- Http Status Code 409 maybe is a better option to this scenario

ricardotulio committed:
Refact CreateUserHandler and change the status code when user already exists

ricardotulio committed:
Fix CreateUserTest
